### PR TITLE
Add Central side SensorEvent hash

### PR DESCRIPTION
**Pull request by connorgorman**
## Description

Add the same sensor event hash that is on the sensor side to Central, which will help when Sensor restarts. For backwards compatibility, this will be computed on Central if it does not exist from an older Sensor.

There is a bug in sensor for roles and role bindings so those aren't adequately deduped, but I'll clean that up in a follow up

Diff between what Central has processed before and after a sensor bounce. Lower change is good
```
sensor_event_queue Operation=Add Type=ClusterStatusUpdate (old: 6.00, new 8.00): change: 33.3333%
sensor_event_queue Operation=Add Type=ImageIntegration (old: 3.00, new 4.00): change: 33.3333%
sensor_event_queue Operation=Add Type=Deployment (old: 175.00, new 217.00): change: 24.0000%
sensor_event_queue Operation=Add Type=Event (old: 2394.00, new 2932.00): change: 22.4728%
sensor_event_queue Operation=Add Type=NetworkFlowUpdate (old: 13.00, new 14.00): change: 7.6923%
sensor_event_queue Operation=Add Type=ClusterHealthInfo (old: 43.00, new 46.00): change: 6.9767%
sensor_event_queue Operation=Add Type=Pod (old: 71.00, new 74.00): change: 4.2254%
sensor_event_queue Operation=Add Type=ProcessIndicator (old: 537.00, new 538.00): change: 0.1862%
sensor_event_queue Operation=Add Type=ServiceAccount (old: 60.00, new 60.00): change: 0.0000%
sensor_event_queue Operation=Add Type=Secret (old: 18.00, new 18.00): change: 0.0000%
sensor_event_queue Operation=Add Type=Role (old: 137.00, new 137.00): change: 0.0000%
sensor_event_queue Operation=Add Type=Node (old: 3.00, new 3.00): change: 0.0000%
sensor_event_queue Operation=Add Type=NetworkPolicy (old: 10.00, new 10.00): change: 0.0000%
sensor_event_queue Operation=Add Type=Namespace (old: 5.00, new 5.00): change: 0.0000%
sensor_event_queue Operation=Add Type=ClusterMetrics (old: 4.00, new 4.00): change: 0.0000%
sensor_event_queue Operation=Add Type=Binding (old: 513.00, new 513.00): change: 0.0000%
sensor_event_queue Operation=Add Type=AlertResults (old: 49.00, new 49.00): change: 0.0000%
```

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Will post the stats here once I get it built and deployed. Also, going to add some metrics from scale testing

### CODE REVIEW

File: central/sensor/service/connection/deduper.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		})
+		utils.Should(err)
+
+		receivedHash = hashValue
+	}
+	prevValue, ok := d.lastReceived[key]
+	if ok && prevValue == receivedHash {
+		return true
+	}
+	d.lastReceived[key] = receivedHash
```
Comment (mtodor) [inline central/sensor/service/connection/deduper.go:79]: Should we consider how much memory this map (set of maps) could use? It's probably less impactful on the sensor side, but on the central side, we have to consider that we will keep copies of data from all sensors.

I'm not aware of in which cases we get a lot of duplicates. If that is in a short period of time, maybe we could consider using some kind of LRU for this purpose? We already use `github.com/hashicorp/golang-lru/v2` in our project.

**Review by mtodor — COMMENTED** [review]
It looks good.

My main concern is memory consumption. It is also important for ACSCS, where we can expect more cases of 1 central <- N secured clusters than on-prem.

File: proto/internalapi/central/sensor_events.proto [hunk 2]
```diff
@@ -55,9 +55,12 @@ message Timing {
 // Next tag: 26.
 message SensorEvent {
     // These fields may be duplicated in the individual events, but avoid the need to branch all the time
-    string         id                           = 1;
-    ResourceAction action                       = 2;
-    Timing         timing                       = 3;
+    string          id                           = 1;
+    ResourceAction  action                       = 2;
+    Timing          timing                       = 3;
+    oneof sensor_hash_oneof {
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply proto/internalapi/central/sensor_events.proto:61]: Yeah sensor hash == 0 is a valid hash value so we need to tell if it's unset

File: central/sensor/service/connection/deduper.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	receivedHash := event.GetSensorHash()
+	// Backwards compatibility with a previous Sensor
+	if event.GetSensorHashOneof() == nil {
+		// Compute the sensor hash
+		d.hasher.Reset()
+		hashValue, err := hashstructure.Hash(event.GetResource(), hashstructure.FormatV2, &hashstructure.HashOptions{
+			TagName: "sensorhash",
+			Hasher:  d.hasher,
+		})
+		utils.Should(err)
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply central/sensor/service/connection/deduper.go:71]: good point, will change

**Comment by connorgorman** [comment]
@mtodor I think I answered your concerns. See slack for more in-depth breakdown of the scale concerns

**Comment by connorgorman** [comment]
@fredrb @lvalerom I added you guys as this has a couple of changes that have sensor<->central implications. Let me know what you think

**Review by fredrb — APPROVED** [review]
Let's also run the compatibility tests just to be sure. [We've had some cases in the past where `jsonpb` was unable to unmarshal proto specs with a new field](https://issues.redhat.com/browse/ROX-14188). But since sensor should not be ahead of central I don't think this will be an issue here.

**Comment by connorgorman** [comment]
@fredrb Good call, note everything in `central` proto should not be serialized using jsonpb and is just serialized proto, but the compatibility tests are good :)

File: sensor/common/deduper/deduper.go [hunk 4]
```diff
@@ -65,15 +59,15 @@ func (d *deduper) Send(msg *central.MsgFromSensor) error {
...
-		return nil
+	hashValue, ok := d.hasher.HashMsg(msg)
+	if ok {
+		// If the hash is valid, then check for deduping
+		if d.lastSent[key] == hashValue {
+			return nil
+		}
+		event.SensorHashOneof = &central.SensorEvent_SensorHash{
+			SensorHash: hashValue,
+		}
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply sensor/common/deduper/deduper.go:70]: Moved it up, thanks

File: proto/internalapi/central/sensor_events.proto [hunk 5]
```diff
@@ -55,9 +55,12 @@ message Timing {
 // Next tag: 26.
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply proto/internalapi/central/sensor_events.proto:55]: done

**Review by mtodor — APPROVED** [review]
Nice work! Looks good.

File: pkg/sensor/hash/hasher.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+func NewHasher() *Hasher {
+	return &Hasher{
+		hasher: fnv.New64a(),
+	}
+}
+
+// HashMsg hashes the message from Sensor
+func (h *Hasher) HashMsg(msg *central.MsgFromSensor) (uint64, bool) {
+	h.hasher.Reset()
+	hashValue, err := hashstructure.Hash(msg.GetEvent().GetResource(), hashstructure.FormatV2, &hashstructure.HashOptions{
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply pkg/sensor/hash/hasher.go:27]: Ah yes, thanks. Meant to do that

**Pull request merged**